src/test/java/BaseTest.java

```diff
@@ -1,11 +1,275 @@
+        actions.doubleClick(playlistElement).perform();
+    }
+
+    //Play song helper functions
+    public void chooseAllSongList() {
+        wait.until(ExpectedConditions.invisibilityOfElementLocated(By.cssSelector(".overlay.loading")));
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("li a.songs"))).click();
+    }
+
+    public void contextClickFirstSong() {
+        WebElement firstSongElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".all-songs tr.song-item:nth-child(1)")));
+        actions.contextClick(firstSongElement).perform();
+    }
+
+    public void choosePlayOption() {
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".song-menu li.playback"))).click();
+    }
+
+    //Hover helper function
+    public WebElement hoverPlay() {
+        WebElement play = driver.findElement(By.cssSelector("[data-testid='play-btn']"));
+        actions.moveToElement(play).perform();
+        return driver.findElement(By.cssSelector("[data-testid='play-btn']"));
+    }
+
+    //Count Songs Helper Functions
+    public void choosePlayListByName(String playlistName){
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//a[contains(text(),'" + playlistName + "')]"))).click();
+    }
+
+    public int countSongs(){
+        return driver.findElements(By.cssSelector("section#playlistWrapper td.title")).size();
+    }
+    public String getPlaylistDetails(){
+        return driver.findElement(By.cssSelector("span.meta.text-secondary span.meta")).getText();
+    }
+
+    public void displayAllSong(){
+        List<WebElement> songList = driver.findElements(By.cssSelector("section#playlistWrapper td.title"));
+        System.out.println("Number of Songs Found:" + countSongs());
+            for (WebElement e: songList){
+                System.out.println(e.getText());
+            }
+    }
+
+    //Helper function for Rename
+    String newPlaylistName = "new name";
+    public void doubleClickPlaylist(){
+        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector(".playlist:nth-child(3)")));
+        actions.doubleClick(playlistElement).perform();
+    }
+
+    public void enterNewPlaylistName() {
+        WebElement playlistInputField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[name='name']")));
+        playlistInputField.sendKeys(Keys.chord(Keys.COMMAND, "A", Keys.BACK_SPACE));
+        playlistInputField.sendKeys(newPlaylistName);
+        playlistInputField.sendKeys(Keys.ENTER);
+    }
+
+    public  boolean doesPlaylistExist(){
+        WebElement playlistElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//a[text()='" + newPlaylistName + "']")));
+        return playlistElement.isDisplayed();
+    }
```

Comment:
1. **BasePage and BaseTest** 

**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

![image](https://user-images.githubusercontent.com/104761482/233972627-2d7ffdc7-6a2f-4e9f-8a3a-a1b4a7a47bc4.png)


**BasePage**  — class serves as a foundation or parent class for individual page classes. It contains common methods and functionalities that are shared across multiple pages in an application.

_Example without a Base Page Class_

![image](https://user-images.githubusercontent.com/104761482/236891493-7f7013c7-fde2-4bf9-9084-3aa7483df8e3.png)


2. **Test in Page Object Model (POM)**

![Clean_Test](https://user-images.githubusercontent.com/104761482/236863813-f30d7bfc-e89d-4419-bdea-4b1517155e39.png)

______________________________________________________________________________________________________________________

For example

The methods clickAvatarIcon() must be stored in the `class LoginPage because the avatar is visible on the web site -login page

![image](https://github.com/SupattraTang/A47-QA-Automation-Framework/assets/104761482/563bab08-fdde-4e02-b50f-d825e56337b5)

**Move all methods to matching classes ( HomePage and LoginPage)**